006-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,95 @@
+# TLS encrypting the Kafka Connect REST API
+
+This proposal reduces access to the Kafka Connect REST API by enabling TLS encryption on the Kafka Connect REST listener when the connectors running on the Kafka Connect cluster are managed by the `KafkaConnector` or `KafkaMirrroMaker2` operators.
+
+## Current situation
+
+Currently, instances of Kafka Connect that are deployed by the Strimzi operators are configured with the default REST API endpoint settings.
+This means that the Kafka Connect REST API endpoint uses HTTP on port 8083 and that the Strimzi `KafkaConnect`, `KafkaConnector` and `KafkaMirrorMaker2` operators make unsecured REST client calls based on this default configuration.
+
+The default network policies created by the Strimzi operators restrict incoming REST API calls to only allow access from the operator pod.
+Users can define further network policies can be provided to override the default policy and allow wider access.
+
+## Motivation
+
+There is currently no way to TLS encrypt the Kafka Connect REST API, leaving the Connect REST listener as perhaps the last key endpoint that cannot currently be restricted using Strimzi.
+
+## Proposal
+
+This proposal changes the behaviour of the `KafkaMirrorMaker2` operator and the `KafkaConnect` operator.
+In the `KafkaConnect` operator case, the behaviour is only changed when the `strimzi.io/use-connector-resources: "true"` annotation is applied in the `KafkaConnect` CR (meaning that all connectors running on the Connect cluster are managed by the `KafkaConnector` operator).
+
+The default behaviour of the `KafkaMirrorMaker2` operator and the `KafkaConnect` operator will be to deploy Kafka Connect clusters with a single TLS encrypted REST API listener.
+A self-signed TLS certificate and key required to enable TLS encryption on the REST API listener are generated by the operator, stored in secrets for use by the operator clients and the Kafka Connect pods.
+The certificates and keys are mounted in each Kafka Connect pod, which create an SSL keystore from the certificate and key and set the following properties in the generated connect configuration to enable an HTTPS listener:
+
+```
+listeners: https://:8443
+rest.advertised.listener: https
+rest.advertised.port: 8443
+listeners.https.ssl.client.auth: none
+listeners.https.ssl.keystore.location: /tmp/kafka/kafka-connect-rest.keystore.p12
+listeners.https.ssl.keystore.password: ***generated password***
+listeners.https.ssl.keystore.type: PKCS12
```

Comment:
IIRC, when Connect is running with multiple replicas, it is using the REST interface for communication between workers and their leader node (basically, requests to create connector etc. are forwarded tot he leader). So, since using self-signed certificate, wouldn't you need to also specify a truststore with the public key to allow the workes to trust the leader when forwarding the request?

Reply:
The truststore was there before but I didn't know about this internal communication in a Kafka Connect cluster. If it's the case then we need the truststore. Sorry @ajborley

Reply:
No worries at all - I had forgotten about the distributed case. I have updated to add the truststore settings back in.